File: translated/f-to-c/DRB079-taskdep3-orig-no.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//tasks with depend clauses to ensure execution order, no data races.

#include <omp.h>
#include <stdio.h>
#include <unistd.h>

int main()
{
    int i, j, k;
    i = 0;

    #pragma omp parallel
    {
        #pragma omp single
        {
            #pragma omp task depend(out:i)
            {
                sleep(3);
                i = 1;
            }
            
            #pragma omp task depend(in:i)
            {
                j = i;
            }
            
            #pragma omp task depend(in:i)
            {
                k = i;
            }
        }
    }

    printf("j = %d  k = %d\n", j, k);

    if (j != 1 && k != 1) {
        printf("Race Condition\n");
    }

    return 0;
} 

File: translated/f-to-go-automodel-bad/DRB079-taskdep3-orig-no.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//tasks with depend clauses to ensure execution order, no data races.

package main

import (
	"fmt"
	"sync"
	"time"
)

func main() {
	var i, j, k int
	i = 0

	//$omp parallel
	var wgParallel sync.WaitGroup
	wgParallel.Add(1)
	go func() {
		defer wgParallel.Done()
		//$omp single
		var wgSingle sync.WaitGroup
		wgSingle.Add(3)

		//$omp task depend (out:i)
		go func() {
			defer wgSingle.Done()
			time.Sleep(3 * time.Second)
			i = 1
		}()

		//$omp task depend (in:i)
		go func() {
			defer wgSingle.Done()
			j = i
		}()

		//$omp task depend (in:i)
		go func() {
			defer wgSingle.Done()
			k = i
		}()

		wgSingle.Wait()
		//$omp end single
	}()
	wgParallel.Wait()
	//$omp end parallel

	fmt.Printf("j =%3d  k =%3d\n", j, k)

	if j != 1 && k != 1 {
		fmt.Println("Race Condition")
	}
}
